coqlspclient/proof_file.py

```diff
@@ -297,107 +313,152 @@ def traverse_expr(expr):
         # HACK: We ignore proofs inside a Module Type since they can't be used outside
         # and should be overriden.
-        if self.coq_file.in_proof and len(self.coq_file.curr_module_type) == 0:
+        if self.in_proof and len(self.curr_module_type) == 0:
             steps = self.__get_steps(proofs)
-            if steps is not None:
-                proofs.append((term, statement_context, steps))
+            proofs.append((term, statement_context, steps))
 
-    def __get_proofs(self) -> List[ProofTerm]:
-        def call_context(calls: List[Tuple]):
-            context, calls = [], [call[0](*call[1:]) for call in calls]
-            [context.append(call) for call in calls if call not in context]
-            return list(filter(lambda term: term is not None, context))
+    def __call_context(self, calls: List[Tuple]):
+        context, calls = [], [call[0](*call[1:]) for call in calls]
+        [context.append(call) for call in calls if call not in context]
+        return list(filter(lambda term: term is not None, context))
 
+    def __get_proofs(self) -> List[ProofTerm]:
         def get_proof_step(step: Tuple[Step, Optional[GoalAnswer], List[Tuple]]):
-            return ProofStep(step[0], step[1], call_context(step[2]))
+            return ProofStep(step[0], step[1], self.__call_context(step[2]))
 
         proofs: List[
             Tuple[Term, List[Tuple[Step, Optional[GoalAnswer], List[Tuple]]]]
         ] = []
-        while not self.coq_file.checked:
+        while not self.checked:
             self.__step()
             # Get context from initial statement
             self.__get_proof(proofs)
 
         try:
             self.__aux_file.didOpen()
         except Exception as e:
-            self.coq_file.close()
+            self.close()
             raise e
 
         proof_steps = [
-            (term, call_context(calls), list(map(get_proof_step, steps)))
+            (term, self.__call_context(calls), list(map(get_proof_step, steps)))
             for term, calls, steps in proofs
         ]
         return list(map(lambda t: ProofTerm(*t), proof_steps))
 
+    def __find_step(self, range: Range) -> Optional[Tuple[ProofTerm, int]]:
+        for proof in self.proofs:
+            for i, proof_step in enumerate(proof.steps):
+                if proof_step.ast.range == range:
+                    return (proof, i)
+        else:
+            return None
+
+    def __find_prev(self, range: Range) -> Tuple[ProofTerm, Optional[int]]:
+        optional = self.__find_step(range)
+        if optional is None:
+            # Previous step may be the definition of the proof
+            for proof in self.proofs:
+                if proof.ast.range == range:
+                    return (proof, -1)
+
+            raise NotImplementedError(
+                "Adding steps outside of a proof is not implemented yet"
+            )
+        else:
+            return optional
```

Comment:
The identation looks nicer like this.
```suggestion
        if optional is not None:
            return optional
            
        # Previous step may be the definition of the proof
        for proof in self.proofs:
            if proof.ast.range == range:
                return (proof, -1)

        raise NotImplementedError(
            "Adding steps outside of a proof is not implemented yet"
        )
```